src/policy/ephemeral_policy.cpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include<policy/ephemeral_policy.h>
+#include<policy/policy.h>
+
+bool CheckValidEphemeralTx(const CTransaction& tx, CFeeRate dust_relay_fee, CAmount txfee, TxValidationState& state)
+{
+    bool has_dust = false;
+    for (const CTxOut& txout : tx.vout) {
+        if (IsDust(txout, dust_relay_fee)) {
+            // We only allow a single dusty output
+            if (has_dust) {
+                return state.Invalid(TxValidationResult::TX_NOT_STANDARD, "dust");
+            }
+            has_dust = true;
+         }
+    }
+
+    // No dust; it's complete standard already
+    if (!has_dust) return true;
+
+    // We never want to give incentives to mine this alone
+    if (txfee != 0) {
+        return state.Invalid(TxValidationResult::TX_NOT_STANDARD, "dust");
+    }
+
+    return true;
+}
+
+std::optional<uint256> CheckEphemeralSpends(const Package& package, CFeeRate dust_relay_rate)
+{
+    // Package is topologically sorted, and PreChecks ensures that
+    // there is up to one dust output per tx.
+
+    assert(std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx != nullptr;}));
+
+    // Running tally of unspent dust
+    std::unordered_set<COutPoint, SaltedOutpointHasher> unspent_dust;
+
+    // If a parent tx has dust, we have to check for the spend
+    // Single dust per tx possible
+    std::map<Txid, uint32_t> map_tx_dust;
+
+    for (const auto& tx : package) {
+        std::unordered_set<Txid, SaltedTxidHasher> child_unspent_dust;
+        for (const auto& tx_input : tx->vin) {
+            // Parent tx had dust, child MUST be sweeping it
+            // if it's spending any output from parent
+            if (map_tx_dust.count(tx_input.prevout.hash)) {
+                child_unspent_dust.insert(tx_input.prevout.hash);
+            }
+        }
+
+        // Now that we've built a list of parent txids
+        // that have dust, make sure that all parent's
+        // dust are swept by this same tx
+        for (const auto& tx_input : tx->vin) {
+            const auto& prevout = tx_input.prevout;
+            // Parent tx had dust, child MUST be sweeping it
+            // if it's spending any output from parent
+            if (map_tx_dust.count(prevout.hash) &&
+                map_tx_dust[prevout.hash] == prevout.n) {
+                 child_unspent_dust.erase(prevout.hash);
+            }
+
+            // We want to detect dangling dust too
+            unspent_dust.erase(tx_input.prevout);
+        }
```

Comment:
In 872c06dacde6c45c2add69c8c15c94571b08119d "policy: Allow dust in transactions, spent in-mempool"

It's not clear to me why there is a rule that a descendant must also spend the dust if they spent any other output from an ancestor. This rule does not seem to be mentioned anywhere in the PR description.

ISTM it would be easier and just as effective to check that all dust outputs end up being spent rather than this slightly convoluted check.

Reply:
Imagine three transactions:

TxA, 0-fee with two outputs, one non-dust, one dust
TxB, spends non-dust
TxC, spends dust

All the dust is spent if `TxA+TxB+TxC` is accepted, but the mining template may just pick up `TxA+TxB`  rather than the three "legal configurations:
0) None
1) `TxA+TxB+TxC`
2) `TxA+TxC`

By requiring the child transaction to sweep any dust from the parent txn, we ensure that there is a single child only, and this child is the only transaction possible for bringing fees, or itself being spent by another child, and so on.

Reply:
I can update the documentation reflecting this if you find the explanation compelling.

Reply:
Ah, I see. Any child must spend the dust output, and I see that this is being enforced in the other variant too.

Additional documentation would be appreciated.

Reply:
added some documentation